benchmarks/benchmark_utils.py

```diff
@@ -0,0 +1,297 @@
+from reco_utils.recommender.ncf.ncf_singlenode import NCF
+from reco_utils.recommender.ncf.dataset import Dataset as NCFDataset
+from reco_utils.recommender.surprise.surprise_utils import surprise_trainset_to_df
+from reco_utils.recommender.fastai.fastai_utils import hide_fastai_progress_bar, cartesian_product, score
+from reco_utils.evaluation.spark_evaluation import SparkRatingEvaluation, SparkRankingEvaluation
+from reco_utils.evaluation.python_evaluation import map_at_k, ndcg_at_k, precision_at_k, recall_at_k
+from reco_utils.evaluation.python_evaluation import rmse, mae, rsquared, exp_var
+
+
+TOP_K = 10
+USER_COL = "UserId"
+ITEM_COL = "MovieId"
+RATING_COL = "Rating"
+TIMESTAMP_COL = "Timestamp"
+PREDICTION_COL = "prediction"
+SEED = 77
+
+
+def prepare_training_als(train):
+    schema = StructType(
+    (
+        StructField(USER_COL, IntegerType()),
+        StructField(ITEM_COL, IntegerType()),
+        StructField(RATING_COL, FloatType()),
+        StructField(TIMESTAMP_COL, LongType()),
+    )
+    )
+    spark = start_or_get_spark()
+    return spark.createDataFrame(train, schema)
+
+
+def train_als(params, data):
+    symbol = ALS(**params)
+    with Timer() as t:
+        model = symbol.fit(data)
+    return model, t
+
+
+def prepare_metrics_als(train, test):
+    schema = StructType(
+    (
+        StructField(USER_COL, IntegerType()),
+        StructField(ITEM_COL, IntegerType()),
+        StructField(RATING_COL, FloatType()),
+        StructField(TIMESTAMP_COL, LongType()),
+    )
+    )
+    spark = start_or_get_spark()
+    return prepare_training_als(train), spark.createDataFrame(test, schema)
+
+
+def predict_als(model, test):
+    with Timer() as t:
+        preds = model.transform(test)
+    return preds, t
+
+
+def recommend_k_als(model, test, train):
+    with Timer() as t:
+        # Get the cross join of all user-item pairs and score them.
+        users = train.select(USER_COL).distinct()
+        items = train.select(ITEM_COL).distinct()
+        user_item = users.crossJoin(items)
+        dfs_pred = model.transform(user_item)
+
+        # Remove seen items.
+        dfs_pred_exclude_train = dfs_pred.alias("pred").join(
+            train.alias("train"),
+            (dfs_pred[USER_COL] == train[USER_COL]) & (dfs_pred[ITEM_COL] == train[ITEM_COL]),
+            how='outer'
+        )
+        top_k_scores = dfs_pred_exclude_train.filter(dfs_pred_exclude_train["train." + RATING_COL].isNull()) \
+            .select('pred.' + USER_COL, 'pred.' + ITEM_COL, 'pred.' + PREDICTION_COL)
+    return top_k_scores, t
+
+
+def prepare_training_svd(train):
+    reader = surprise.Reader('ml-100k', rating_scale=(1, 5))
+    return surprise.Dataset.load_from_df(train.drop(TIMESTAMP_COL, axis=1), reader=reader).build_full_trainset()
+
+
+def train_svd(params, data):
+    model = surprise.SVD(**params)
+    with Timer() as t:
+        model.fit(data)
+    return model, t
+
+
+def predict_svd(model, test):

```

Comment:
We have some utilities for doing these with Surprise:
https://github.com/Microsoft/Recommenders/blob/master/reco_utils/recommender/surprise/surprise_utils.py

Reply:
very good catch, fixed in the next commit